widgets/url-sync/index.js

```diff
@@ -0,0 +1,172 @@
+ * it'll use the query parameters using the modern history API.
+ */
+class URLSync {
+  constructor(urlUtils, options) {
+    this.__initLast = true;
+    this.urlUtils = urlUtils;
+    this.originalConfig = null;
+    this.timer = timerMaker(getNow());
+    this.threshold = options.threshold || 700;
+    this.trackedParameters = options.trackedParameters || ['query', 'attribute:*', 'index', 'page'];
+  }
+
+  getConfiguration(currentConfiguration) {
+    this.originalConfig = currentConfiguration;
+    var queryString = this.urlUtils.readUrl();
+    var config = AlgoliaSearchHelper.getConfigurationFromQueryString(queryString);
+    return config;
+  }
+  
+  init(state, helper) {
+    var self = this;
+    this.urlUtils.onpopstate(function() {
+      var qs = self.urlUtils.readUrl();
+      var partialState = AlgoliaSearchHelper.getConfigurationFromQueryString(qs);
+      var fullState = merge({}, self.originalConfig, partialState);
+      //compare with helper.state
+      var partialHelperState = helper.getState(self.trackedParameters);
+      var fullHelperState = merge({}, self.originalConfig, partialHelperState);
+
+      if(isEqual(fullHelperState, fullState)) return;
+
+      forEach(window.document.querySelectorAll('input'), function(i) {
+        i.blur();
+      });
+
+      helper.overrideStateWithoutTriggeringChangeEvent(fullState);
+      helper.search();
+    });
+  }
+
+  render({helper}){
+    var helperState = helper.getState(this.trackedParameters);
+    var currentQueryString = this.urlUtils.readUrl();
+    var foreignConfig = AlgoliaSearchHelper.getForeignConfigurationInQueryString(currentQueryString);
+    var urlState = AlgoliaSearchHelper.getConfigurationFromQueryString(currentQueryString);
+
+    if (isEqual(helperState, urlState)) return;
+
+    var qs = helper.getStateAsQueryString({filters: this.trackedParameters, moreAttributes: foreignConfig});
+    if(this.timer() < this.threshold) {
+      this.urlUtils.replaceState(qs);
+    }
+    else {
+      this.urlUtils.pushState(qs);
+    }
+  }
+}
+
+/**
+ * Instanciate a url sync widget. This widget let you synchronize the search
+ * parameters with the URL. It can operate with legacy API and hash or it can use
+ * the modern history API. By default, it will use the modern API, but if you are
+ * looking for compatibility with IE8 and IE9, then you should set 'useHash' to
+ * true.
+ * @param {number} threshold time in ms after which a new state is created in the browser
+ * history. The default value is 700.
+ * @param {string[]} trackedParameters parameters that will be synchronized in the
+ * URL. By default, it will track the query, all the refinable attribute (facets and numeric
+ * filters), the index and the page.
+ * @param {boolean} useHash if set to true, the url will be hash based. Otherwise,
+ * it'll use the query parameters using the modern history API.
+ * @return {object} the widget instance
+ */
+function urlSync(params) {
+  var options = params || {};
```

Comment:
You can use default params and destructuring like in other widgets for this usage.